sdk/clientcore/http-netty4/src/main/java/io/clientcore/http/netty4/implementation/CoreHttpProxyHandler.java

```diff
@@ -0,0 +1,256 @@
+            innerHeaders = response.headers();
+
+            if (response.status().code() == 407) {
+                /*
+                 * Attempt to CONNECT to the proxy resulted in a request for authentication, parse the
+                 * Proxy-Authenticate headers returned for challenges and update the ChallengeHolder reference. This
+                 * will allow subsequent requests to CONNECT to this proxy to use the challenges returned to create a
+                 * Proxy-Authorization header.
+                 */
+                proxyChallengeHolder.set(extractChallengesFromHeaders(response.headers()));
+            } else if (response.status().code() == 200) {
+                /*
+                 * Attempt to CONNECT to the proxy succeeded, retrieve the Proxy-Authorization header passed in the
+                 * Channel's attributes to compare it against a potential Proxy-Authentication-Info response header.
+                 * This header is used by the server to validate to the client that it received the correct request.
+                 */
+                validateProxyAuthenticationInfo(
+                    NettyUtility.get(response.headers(), PROXY_AUTHENTICATION_INFO, PROXY_AUTHENTICATION_INFO_NAME),
+                    lastUsedAuthorizationHeader);
+            }
+        }
+
+        boolean responseComplete = o instanceof LastHttpContent;
+        if (responseComplete) {
+            if (status == null) {
+                throw new HttpProxyHandler.HttpProxyConnectException(exceptionMessage("missing response"),
+                    innerHeaders);
+            } else if (status.code() != 200) {
+                throw new HttpProxyHandler.HttpProxyConnectException(exceptionMessage("status: " + status),
+                    innerHeaders);
+            }
+        }
+
+        return responseComplete;
+    }
+
+    /*
+     * Search the response HttpHeaders for Proxy-Authenticate headers.
+     */
+    private static List<AuthenticateChallenge> extractChallengesFromHeaders(HttpHeaders headers) {
+        List<String> proxyAuthenticateValues
+            = NettyUtility.getAll(headers, HttpHeaderNames.PROXY_AUTHENTICATE, HttpHeaderName.PROXY_AUTHENTICATE);
+
+        List<AuthenticateChallenge> challenges = new ArrayList<>();
+        for (String proxyAuthenticateValue : proxyAuthenticateValues) {
+            challenges.addAll(AuthUtils.parseAuthenticateHeader(proxyAuthenticateValue));
+        }
+
+        return challenges;
+    }
+
+    /*
+     * Validate the Proxy-Authorization header used in authentication against the server's Proxy-Authentication-Info
+     * header. This header is an optional return by the server so this validation is guaranteed to happen.
+     */
+    private void validateProxyAuthenticationInfo(String infoHeader, String authorizationHeader) {
+        // Server didn't return a 'Proxy-Authentication-Info' header, nothing to consume.
+        if (CoreUtils.isNullOrEmpty(infoHeader)) {
```

Comment:
why doesn't `authorizationHeader` need a check?

Reply:
Hmm, it shouldn't ever be null here, but doesn't hurt to validate that.